lab_4_summarization_textrank/main.py

```diff
@@ -0,0 +1,302 @@
+
+    def encode_sentences(self, sentences: tuple[Sentence, ...]) -> None:
+        """
+        Enriches the instances of sentences with their encoded versions
+        :param sentences: a sequence of sentences
+        :return: a list of sentences with their preprocessed versions
+        """
+        pass
+
+
+def calculate_similarity(sequence: Union[list, tuple], other_sequence: Union[list, tuple]) -> float:
+    """
+    Calculates similarity between two sequences using Jaccard index
+    :param sequence: a sequence of items
+    :param other_sequence: a sequence of items
+    :return: similarity score
+    """
+    pass
+
+
+class SimilarityMatrix:
+    """
+    A class to represent relations between sentences
+    """
+
+    _matrix: list[list[float]]
+
+    def __init__(self) -> None:
+        """
+        Constructs necessary attributes
+        """
+        pass
+
+    def get_vertices(self) -> tuple[Sentence, ...]:
+        """
+        Returns a sequence of all vertices present in the graph
+        :return: a sequence of vertices
+        """
+        pass
+
+    def calculate_inout_score(self, vertex: Sentence) -> int:
+        """
+        Retrieves a number of vertices that are similar (i.e. have similarity score > 0) to the input one
+        :param vertex
+        :return:
+        """
+        pass
+
+    def add_edge(self, vertex1: Sentence, vertex2: Sentence) -> None:
+        """
+        Adds or overwrites an edge in the graph between the specified vertices
+        :param vertex1:
+        :param vertex2:
+        :return:
+        """
+        pass
+
+    def get_similarity_score(self, sentence: Sentence, other_sentence: Sentence) -> float:
+        """
+        Gets the similarity score for two sentences from the matrix
+        :param sentence
+        :param other_sentence
+        :return: the similarity score
+        """
+        pass
+
+    def fill_from_sentences(self, sentences: tuple[Sentence, ...]) -> None:
+        """
+        Updates graph instance with vertices and edges extracted from sentences
+        :param sentences
+        :return:
+        """
+        pass
+
+
+class TextRankSummarizer:
+    """
+    TextRank for summarization
+    """
+
+    _scores: dict[Sentence, float]
+    _graph: SimilarityMatrix
+
+    def __init__(self, graph: SimilarityMatrix) -> None:
+        """
+        Constructs all the necessary attributes
+        :param graph: the filled instance of the similarity matrix
+        """
+        pass
+
+    def update_vertex_score(
+            self, vertex: Sentence, incidental_vertices: list[Sentence], scores: dict[Sentence, float]
+    ) -> None:
+        """
+        Changes vertex significance score using algorithm-specific formula
+        :param vertex: a sentence
+        :param incidental_vertices: vertices with similarity score > 0 for vertex
+        :param scores: current vertices scores
+        :return:
+        """
+        pass
+
+    def train(self) -> None:
```

Comment:
do not we provide implementation for train as we did in lab 3?

Reply:
No

Reply:
why??

Reply:
because